src/System.Text.JsonLab/System/Text/Json/JsonWriter.cs

```diff
@@ -1537,4 +1462,2816 @@ private int AddNewLineAndIndentation(Span<char> buffer)
+
+            try
+            {
+                if (!isFirst)
+                    byteBuffer[idx++] = JsonConstants.ListSeperator;
+
+                if (prettyPrint)
+                {
+                    while (indent-- >= 0)
+                    {
+                        byteBuffer[idx++] = JsonConstants.Space;
+                        byteBuffer[idx++] = JsonConstants.Space;
+                    }
+                }
+
+                byteBuffer[idx++] = token;
+
+                written = idx;
+                return true;
+            }
+            catch (IndexOutOfRangeException)
+            {
+                written = default;
+                return false;
+            }
+        }
+
+        public static bool TryWriteEndUtf8(Span<byte> byteBuffer, byte token, out int written, int indent, bool isFirst, bool prettyPrint = false)
+        {
+            int idx = 0;
+            try
+            {
+                if (prettyPrint)
+                    idx += AddNewLineAndIndentation(byteBuffer, indent);
+
+                byteBuffer[idx++] = token;
+
+                written = idx;
+                return true;
+            }
+            catch (IndexOutOfRangeException)
+            {
+                written = default;
+                return false;
+            }
+        }
+    }
+
+    struct Customer : IWritable
+    {
+        //readonly string _name;
+        readonly long _value;
+        readonly int _indent;
+        readonly bool _isFirst;
+        readonly bool _prettyPrint;
+
+        /*public Customer(string name, int[] value, int indent, bool isFirst, bool prettyPrint)
+        {
+            _name = name;
+            _value = value;
+            _indent = indent;
+            _isFirst = isFirst;
+            _prettyPrint = prettyPrint;
+        }*/
+
+        /*public Customer(string name, string value, int indent, bool isFirst, bool prettyPrint)
+        {
+            _name = name;
+            _value = value;
+            _indent = indent;
+            _isFirst = isFirst;
+            _prettyPrint = prettyPrint;
+        }*/
+
+        public Customer(long value, int indent, bool isFirst, bool prettyPrint)
+        {
+            _value = value;
+            _indent = indent;
+            _isFirst = isFirst;
+            _prettyPrint = prettyPrint;
+        }
+
+        public bool TryWrite(Span<byte> buffer, out int written, StandardFormat format = default)
+        {
+            //return JsonWriterHelper.TryWriteValueUtf8(buffer, MemoryMarshal.AsBytes(_name.AsSpan()), MemoryMarshal.AsBytes(_value.AsSpan()), out written, _indent, _isFirst, _prettyPrint);
+            //return JsonWriterHelper.TryWriteValueUtf8(buffer, MemoryMarshal.AsBytes(_name.AsSpan()), _value, out written, _indent, _isFirst, _prettyPrint);
+            return JsonWriterHelper.TryWriteValueUtf8(buffer, _value, out written, _indent, _isFirst, _prettyPrint);
+        }
+    }
+
+    struct CustomerArray : IWritable
```

Comment:
@KrzysztofCwalina, is this the type of implementation you had in mind (the TryWrite method ends up calling this helper https://github.com/dotnet/corefxlab/pull/2369/files#diff-035eb68e9f76877e164e42073ac9ded2R1588)?

Reply:
Yeah, this is what I had in mind. I would not call the arguments you pass in "internal state". If somebody designed just such static methods for JSON writing scenarios (i.e. not having writer type with instance methods), they would probably need these parameters.

Having said that, it seems like the wins from IWritable are not super clear.

Reply:
Presumably some percentage of this is from struct and parameter copies. Wonder how much? If `TryWriteBytes<T>`  took `ref T value`, what sort of difference would that make? If we pass state by ref how much further would that get us?

Here one way to pass state by ref. Not pretty, but unsure if there is another way:

``` C#
public ref struct Foo
{
    private struct State
    {
        public readonly bool PrettyPrint;
        public int Indent;
        public bool IsFirst;
    }

    private State _state;

    public Foo(int indent)
    {
        _state = new State { Indent = indent };
    }

    public int Indent { get => _state.Indent; set => _state.Indent = value; }

    public unsafe void WriteStuff()
    {
        fixed(State* state = &_state)
        {
            (new WritableWrapper(state)).TryWrite(Span<byte>.Empty, out int _);
        }
    }

    private readonly unsafe struct WritableWrapper : IWritable
    {
        readonly State* _state;

        public WritableWrapper(State* state)
        {
            _state = state;
        }

        public bool TryWrite(Span<byte> buffer, out int written, StandardFormat format = default)
        {
            Console.WriteLine($"Indent: {_state->Indent}");
            written = 0;
            return true;
        }
    }
}

```  

---
In reply to: [197203393](https://github.com/dotnet/corefxlab/pull/2369#discussion_r197203393) [](ancestors = 197203393)